report/data_tasks/report/create_from_scratch/99_hubspot_write/02_contact_property_update/01_create_view.sql

```diff
@@ -2,18 +2,64 @@ DROP VIEW IF EXISTS hubspot.contact_property_update CASCADE;
 CREATE VIEW hubspot.contact_property_update AS (
+    WITH contact_users AS (
+        SELECT
+            contacts.id AS hs_object_id,
+            -- This is a little confusing, but the id here comes from H originally,
+            -- but is just processed through the LMS reporting system
+            MAX(users.id) AS h_user_id,
+            MIN(users.registered_date) AS lms_first_login
+        FROM hubspot.contacts
+        -- We are going to assume people never _stop_ being a teacher. This means
+        -- we will never remove the `Teacher` status from a contact. The upside is
+        -- we will update far fewer records, which should happen quicker.
+        JOIN lms.users ON
+            -- Emails in the contacts table are already stored as lower case
+            contacts.email = LOWER(users.email)
+            -- We definitely want to NOT join onto student information to keep that
+            -- out of Hubspot
+            AND users.is_teacher = True
+        GROUP BY contacts.id
+    )
+
     SELECT
-        contacts.id AS hs_object_id,
-        'Teacher' AS lms_role
-    FROM hubspot.contacts
-    -- We are going to assume people never _stop_ being a teacher. This means
-    -- we will never remove the `Teacher` status from a contact. The upside is
-    -- we will update far fewer records, which should happen quicker.
-    JOIN lms.users ON
-        -- Emails in the contacts table are already stored as lower case
-        contacts.email = LOWER(users.email)
-        -- We definitely want to NOT join onto student information to keep that
-        -- out of Hubspot
-        AND users.is_teacher = True
-    GROUP BY contacts.id
+        contact_users.hs_object_id,
+        contact_users.h_user_id,
+        'Teacher' AS lms_role,
+        contact_users.lms_first_login,
+        last_login.lms_last_login,
+        report.present_date('semester', CURRENT_DATE) AS lms_reporting_semester,
+        COALESCE(course_activity.lms_annotations_this_semester, 0) AS lms_annotations_this_semester,
+        COALESCE(course_activity.lms_active_courses_this_semester, 0) AS lms_active_courses_this_semester
+    FROM contact_users
+
+    LEFT OUTER JOIN (
+        SELECT
+            user_id,
+            SUM(annotation_count) AS lms_annotations_this_semester,
+            COUNT(DISTINCT(group_roles.group_id)) AS lms_active_courses_this_semester
+        FROM lms.users
+        JOIN lms.group_roles ON
+            group_roles.user_id = users.id
+            AND group_roles.role = 'teacher'
+        JOIN lms.group_bubbled_activity ON
+            group_bubbled_activity.group_id = group_roles.group_id
+            AND group_bubbled_activity.created_week >= report.multi_truncate('semester', CURRENT_DATE)
+        JOIN lms.groups ON
+            groups.id = group_roles.group_id
+        GROUP BY user_id
+    ) AS course_activity ON
+        course_activity.user_id = contact_users.h_user_id
```

Comment:
I've swapped the normal way of doing this CTE's around this time because CTE's are always executed separately.

If we calculate these rows as CTE's we'd calculate many rows we don't need to. My untested hope is Postgres can use the join to only calculate required rows.